File: application/services/diff_payload.py
"""Normalize unified diff text into a reusable payload shape."""
from __future__ import annotations

from typing import Any, Dict, List, Optional


def build_diff_payload(patch: str) -> Dict[str, Any]:
    text = patch or ""
    lines: List[Dict[str, Any]] = []
    files: List[Dict[str, Any]] = []
    current_file: Optional[str] = None
    additions = 0
    deletions = 0

    for raw in text.splitlines():
        if raw.startswith("+++ b/"):
            current_file = raw[6:]
            if current_file not in {item["path"] for item in files}:
                files.append({"path": current_file})
            lines.append({"type": "meta", "content": raw, "path": current_file})
            continue
        if raw.startswith("--- a/"):
            lines.append({"type": "meta", "content": raw, "path": current_file})
            continue
        if raw.startswith("@@") or raw.startswith("diff --git"):
            lines.append({"type": "meta", "content": raw, "path": current_file})
            continue
        if raw.startswith("+") and not raw.startswith("+++"):
            additions += 1
            lines.append({"type": "add", "content": raw[1:], "path": current_file})
            continue
        if raw.startswith("-") and not raw.startswith("---"):
            deletions += 1
            lines.append({"type": "remove", "content": raw[1:], "path": current_file})
            continue
        lines.append({"type": "context", "content": raw, "path": current_file})

    return {
        "raw_patch": text,
        "files": files,
        "stats": {
            "files_changed": len(files),
            "additions": additions,
            "deletions": deletions,
            "lines_changed": additions + deletions,
        },
        "lines": lines,
    }

File: ui/src/utils/diffPayload.js
export function buildDiffPayload(patch = '') {
  const text = patch || '';
  if (!text) {
    return {
      raw_patch: '',
      files: [],
      stats: {
        files_changed: 0,
        additions: 0,
        deletions: 0,
        lines_changed: 0,
      },
      lines: [],
    };
  }
  const files = [];
  const lines = [];
  let currentFile = null;
  let additions = 0;
  let deletions = 0;

  text.split('\n').forEach((raw) => {
    if (raw.startsWith('+++ b/')) {
      currentFile = raw.slice(6);
      if (!files.some((item) => item.path === currentFile)) {
        files.push({ path: currentFile });
      }
      lines.push({ type: 'meta', content: raw, path: currentFile });
      return;
    }
    if (raw.startsWith('--- a/')) {
      lines.push({ type: 'meta', content: raw, path: currentFile });
      return;
    }
    if (raw.startsWith('@@') || raw.startsWith('diff --git')) {
      lines.push({ type: 'meta', content: raw, path: currentFile });
      return;
    }
    if (raw.startsWith('+') && !raw.startsWith('+++')) {
      additions += 1;
      lines.push({ type: 'add', content: raw.slice(1), path: currentFile });
      return;
    }
    if (raw.startsWith('-') && !raw.startsWith('---')) {
      deletions += 1;
      lines.push({ type: 'remove', content: raw.slice(1), path: currentFile });
      return;
    }
    lines.push({ type: 'context', content: raw, path: currentFile });
  });

  return {
    raw_patch: text,
    files,
    stats: {
      files_changed: files.length,
      additions,
      deletions,
      lines_changed: additions + deletions,
    },
    lines,
  };
}
